cl/lib/utils.py

```diff
@@ -118,3 +118,75 @@ def human_sort(
         sorter = lambda item: [convert(c) for c in re.split("([0-9]+)", item)]
 
     return sorted(unordered_list, key=sorter)
+
+
+def microservice(
+    service: str,
+    method: str = "POST",
+    doc: Any = None,
+    file: bytes = None,
+    filename: str = None,
+    filepath: str = None,
+    data=None,
+    params=None,
+) -> Response:
+    """Call a Microservice endpoint
+
+    This is a helper utility to call our microservices.  To see a list of Endpoints
+    check out the settings file cl/settings/public.py.
+
+    Because of the various ways our db is setup we have a few different params we use
+    in this function.
+
+    :param service: The service to call
+    :param method: The method to use (defaults to POST)
+    :param doc: The document as a db object
+    :param file: The file as a byte array
+    :param filename: The filename of the file
+    :param filepath: The filepath of the file
+    :param data: The data to send
+    :param params: The params to send
+    :return: The response from the microservice
+    """
+    from cl.audio.models import Audio
+    from cl.search.models import Opinion, RECAPDocument
+
+    services = settings.MICROSERVICE_URLS
+
+    req = Request(
+        method=method,
+        url=services[service]["url"],  # type: ignore
+    )
+    # Add file from filepath
+    if filepath:
+        with open(filepath, "rb") as f:
+            req.files = {"file": (filepath, f.read())}
+
+    # Handle our documents based on the type of model object
+    # Sadly these are not uniform
+    if doc and type(doc) == RECAPDocument:
+        req.files = {
+            "file": (doc.filepath_local.name, doc.filepath_local.read())
+        }
+    if doc and type(doc) == Opinion:
+        req.files = {"file": (doc.local_path.name, doc.local_path.read())}
+    if doc and type(doc) == Audio:
+        req.files = {
+            "file": (
+                doc.local_path_original_file.name,
+                doc.local_path_original_file.read(),
+            )
+        }
```

Comment:
I'm no fan of nesting if's, but I'd put this all under one `if doc` statement, with nested `if type(doc) == foo` statements inside.

Reply:
noted.

Reply:
Two more ideas come to mind here now that the if is indented.

First, I *think* the design is that we'd only ever get file, filepath, or item. An assertion like this would ensure that nobody screwed that up and would clarify use of the function:

```
assert sum([bool(file), bool(filepath), bool(item)]) == 1, \
    "Only one of file, filepath, or item should be passed."
```

Second, does it make sense to move file and filepath next to each other and put the three approaches into an if/else? 

```
if file:
    ...
elif filepath:
    ...
elif item:
    ...
else:
    raise ValueError("file, filepath, or item must be provided")
```